### DISCUSSION TITLE
Finding Anomalies Using STUMPI

### DISCUSSION THREAD

**Discussion by seanlaw** [TDAmeritrade/stumpy#212] [Q&A]
@hmcoservit originally [wrote](https://github.com/TDAmeritrade/stumpy/issues/207#issuecomment-653441736): 

> @seanlaw According to my tests so far on anomaly detection, I can see the effect of Z-normalization more and more. It really makes it difficult to detect the peaks in time series (using STUMPI left MP). Also when sub sequences are flat (but noisy) the matrix profile would not give ideal results for anomaly discovery specially when the sub sequence value is small. Therefore, impacting the precision heavily. What do you think about this? I am not sure if it is possible to adapt it to suit anomaly detection more.

**Comment by seanlaw** [comment]
@hmcoservit Moving the discussion here as the [other issue](https://github.com/TDAmeritrade/stumpy/issues/207#issuecomment-653441736) is focused on implementing non-normalized distances. 


So, since matrix profiles mainly look at subsequences, if you are looking for (roughly) single point peaks in a time series then matrix profiles are not a good approach. Have you ever looked at:

[Twitter’s Anomaly Detection ](https://blog.twitter.com/engineering/en_us/a/2015/introducing-practical-and-robust-anomaly-detection-in-a-time-series.html)

Or

[Netflix Outlier Detection](https://link.medium.com/dtpwLGE3O7)

Perhaps, you can provide some examples that we can explore together? I am sure that this will help others as well

**Comment by hmcoservit** [comment]
@seanlaw excluding contextual anomalies in time series, the target is to detect collective and point anomalies. 

The effect of z-normalization might be more visible when either the sub sequence value is too small or there are flat regions in the time series as  [discussed in this paper](https://www.researchgate.net/publication/338811628_Implications_of_Z-Normalization_in_the_Matrix_Profile). They propose a new solution to overcome this in the case of anomaly detection. Another thing that can be done to overcome this is to increase the sub sequence length to a long period, but doing so would eventually add delays to detection as new points coming in do not update the MP values right away. 

Regarding Netflix's solution, is is mainly for processing data in batches, and unsuitable for real-time streaming applications [according to this article ](https://www.sciencedirect.com/science/article/pii/S0925231217309864). And for Twitter's one, it doesn't seem to be maintained and has been known to [perform worse than RRCF ](https://github.com/numenta/NAB) according to NAB. 

I will try to share some visual examples here as soon as possible ;)

**Comment by seanlaw** [comment]
@hmcoservit I'm going to close this for now but please feel free to re-open when you have some examples to share

**Comment by hmcoservit** [comment]
@seanlaw Thanks Sean, for sure.

**Comment by hmcoservit** [comment]
@seanlaw let me share a few examples here (simulating data as a stream), it would be great to have your opinion:

- The data is from NAB, can be[ accessed here ](https://github.com/numenta/NAB/blob/master/data/realAWSCloudwatch/ec2_cpu_utilization_825cc2.csv)
- Confidence area is calculated using the 3-sigma rule of the previous 400 matrix profile values (example for anomaly detection)
- Matrix profile is shifted forward to align it with the current seen (processed) sub-sequence 
- Matrix profile is calculated incrementally using Stumpi and the LEFT_MP is displayed (updating the mp only on past subsequences)
- Plots are displayed for different m values (the dataset length is 4032 points)

It can be seen that the anomaly region is not evident in most of the matrix profile value probably due to Z-normalization. And finding robust parameters for m and moving average is difficult. I am very curious to see if [aampi ](https://github.com/TDAmeritrade/stumpy/issues/221) can better detect the start of collective anomalies or point anomalies. 

![newplot - 2020-08-19T160951 202](https://user-images.githubusercontent.com/60774604/90645607-71962c80-e236-11ea-97a6-e5848c1d3ccd.png)
![newplot - 2020-08-19T161034 638](https://user-images.githubusercontent.com/60774604/90645673-896db080-e236-11ea-8d39-53a351b114c0.png)
![newplot - 2020-08-19T161119 416](https://user-images.githubusercontent.com/60774604/90645782-a904d900-e236-11ea-8911-891ba551221a.png)
![newplot - 2020-08-19T161203 123](https://user-images.githubusercontent.com/60774604/90645836-bae67c00-e236-11ea-8434-4736cd5a719e.png)
![newplot - 2020-08-19T161256 613](https://user-images.githubusercontent.com/60774604/90645958-e4070c80-e236-11ea-98e8-bb2d7a3ca9fd.png)
![newplot - 2020-08-19T161352 463](https://user-images.githubusercontent.com/60774604/90646016-fd0fbd80-e236-11ea-815b-a37f0ed37be4.png)

**Comment by seanlaw** [comment]
@hmcoservit Thank you for the context. I took at stab at implementing `aampi` (i.e., `stumpi` but without z-normalization) - [please take a look here](https://github.com/TDAmeritrade/stumpy/issues/221)
